packages/babel-plugin-transform-es2015-for-of/src/index.js

```diff
@@ -89,16 +94,48 @@ export default function ({ messages, template, types: t }) {
     return nodes;
   }
 
+  function replaceWithArray(path) {
+    if (path.parentPath.isLabeledStatement()) {
+      path.parentPath.replaceWithMultiple(_ForOfStatementArray(path));
+      return true;
+    } else {
+      path.replaceWithMultiple(_ForOfStatementArray(path));
+      return true;
+    }
+    return false;
+  }
+
+  function optimize(path, right) {
+    if (right.isArrayExpression() || right.isGenericType("Array")) {
+      return replaceWithArray(path);
+    } else if (right.isIdentifier() && right.isPure()) {
+      const binding = path.scope.getBinding(right.node.name);
+      return optimize(path, binding.path.get("init"));
+    } else if (right.isCallExpression() && (
+      isArrayFrom(right.get("callee").node) ||
```

Comment:
I'd rather see this in https://github.com/babel/babel/blob/622672d397805a22ffd8bf47e750e4a871870d41/packages/babel-traverse/src/path/inference/inferers.js#L148-L150.